runtime/lua/vim/lsp.lua

```diff
@@ -0,0 +1,791 @@
+    "autocmd FileType %s silent lua vim.lsp.attach_to_buffer(0, %d)",
+    table.concat(config.filetypes, ','),
+    config.client_id))
+  return config.client_id
+end
+
+nvim_command("autocmd VimLeavePre * lua vim.lsp.stop_all_clients()")
+
+function lsp.get_client_by_id(client_id)
+  return LSP_CLIENTS[client_id]
+end
+
+function lsp.get_client_by_name(name)
+  local config = LSP_CONFIGS[name]
+  if config.client_id then
+    return LSP_CLIENTS[config.client_id]
+  end
+end
+
+function lsp.stop_client(client_id, force)
+  local client = LSP_CLIENTS[client_id]
+  if client then
+    client.stop(force)
+  end
+end
+
+function lsp.stop_all_clients(force)
+  for client_id, client in pairs(LSP_CLIENTS) do
+    client.stop(force)
+  end
+end
+
+--- Send a request to a server and return the response
+-- @param method [string]: Name of the request method
+-- @param params [table] (optional): Arguments to send to the server
+-- @param bufnr [number] (optional): The number of the buffer
+-- @param filetype [string] (optional): The filetype associated with the server
+-- @param server_name [string] (optional)
+--
+-- @returns: success?, request_id, cancel_fn
+function lsp.buf_request(bufnr, method, params, callback)
+  if callback then
+    assert(type(callback) == 'function', "buf_request callback must be a function")
+  end
+  local client_request_ids = {}
+  for_each_buffer_client(bufnr, function(client, client_id)
+    local request_callback = callback
+    if not request_callback then
+      request_callback = client.default_server_callbacks[method]
+        or error(string.format("buf_request callback is empty and no default client was found for client %s", client.name))
+    end
+    local request_success, request_id = client.request(method, params, function(err, result)
+      -- TODO pass client here?
+      request_callback(err, result, client_id)
+    end)
+
+    -- This could only fail if the client shut down in the time since we looked
+    -- it up and we did the request, which should be rare.
+    if request_success then
+      client_request_ids[client_id] = request_id
+    end
+  end)
+
+  local function cancel_request()
+    for client_id, request_id in pairs(client_request_ids) do
+      local client = LSP_CLIENTS[client_id]
+      client.rpc.notify('$/cancelRequest', { id = request_id })
+    end
+  end
+
+  return client_request_ids, cancel_request
+end
+
+--- Send a request to a server, but don't wait for the response
+-- @param method [string]: Name of the request method
+-- @param params [string]: Arguments to send to the server
+-- @param cb [function|string] (optional): Either a function to call or a string to call in vim
+-- @param bufnr [number] (optional): The number of the buffer
+-- @param filetype [string] (optional): The filetype associated with the server
+-- @param server_name [string] (optional)
+--
+-- @returns: The table of request id
+function lsp.buf_request_sync(bufnr, method, params, timeout_ms)
```

Comment:
The result will be passed in the return value since it is synchronous. I forgot to change the docs on that. It's a table of `{[client_id] = request_result}`

Reply:
oops, sorry. I mistaken it as 'reques_async'.

Reply:
Yeah I purposely chose to make the `_sync` case special since it's more atypical and also potentially more faulty. The sleep logic is still something I want to figure out a better solution for.